File: randomPickWithWeight/randomPickWithWeight_lowerbound.java
class Solution {

    public Solution(int[] w) {
        random = new Random();
        preSum = new int[w.length];
        preSum[0] = w[0];
        for (int i = 1; i < w.length; i++) {
            preSum[i] = preSum[i - 1] + w[i];
        }
        root = buildTree(0, w.length - 1);
    }
    
    public int pickIndex() {
        int len = preSum.length;
        if (len == 1) return 0;
        int n = random.nextInt(preSum[len - 1]) + 1;
        TreeNode lb = findLowerBound(n);
        return lb.index;
    }
    
    private TreeNode findLowerBound(int n) {
        TreeNode lb = null;
        TreeNode cur = root;
        while (cur != null) {
            if (preSum[cur.index] < n) {
                cur = cur.right;
            } else {
                lb = cur;
                cur = cur.left;
            }
        }
        return lb;
    }
    
    private TreeNode buildTree(int left, int right) {
        if (left > right) return null;
        int mid = left + (right - left) / 2;
        TreeNode root = new TreeNode(mid);
        root.left = buildTree(left, mid - 1);
        root.right = buildTree(mid + 1, right);
        return root;
    }
    
    private int[] preSum;
    Random random;
    TreeNode root;
}

class TreeNode {
    public TreeNode(int index) { this.index = index; }
    TreeNode left;
    TreeNode right;
    int index;
}

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(w);
 * int param_1 = obj.pickIndex();
 */

File: 528.randomPickWithWeight/randomPickWithWeight_lowerbound.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    class TreeNode {
    public:
        TreeNode(int _key, int _val) : left(nullptr), right(nullptr), key(_key), val(_val){};
        TreeNode *left, *right;
        int key, val;
    };
    Solution(vector<int> w) {
        preSum = w;
        for (int i = 1; i < w.size(); i++)
            preSum[i] += preSum[i - 1];
        root = buildTree(preSum, 0, preSum.size() - 1);
        srand(time(0));
    }
    ~Solution() {
        deleteTree(root);
    }
    
    int pickIndex() {
        int res = -1;
        int len = preSum.size();
        if (len == 0)
            return res;
        int weight = rand() % preSum.back() + 1;
        //cout << "weight: " << weight << endl;
        TreeNode *node = findIndex(root, weight);
        if (node)
            res = node->val;
        return res;
    }
    TreeNode* findIndex(TreeNode *root, int weight) {
        TreeNode *lb = nullptr;
        while (root) {
            if (root->key >= weight) {
                lb = root;
                root = root->left;
            } else {
                root = root->right;
            }
        }
        return lb;
    }
    TreeNode* buildTree(vector<int>& w, int left, int right) {
        if (left > right)
            return nullptr;
        int mid = left + (right - left) / 2;
        TreeNode *node = new TreeNode(w[mid], mid);
        //cout << node->key << "->" << node->val << endl;
        node->left = buildTree(w, left, mid - 1);
        node->right = buildTree(w, mid + 1, right);
        return node;
    }
    void deleteTree(TreeNode* root) {
        if (!root)
            return;
        deleteTree(root->left);
        deleteTree(root->right);
        delete root;
    }
    TreeNode* root;
    vector<int> preSum;
};

int main() {
    vector<int> w = {1, 3};
    Solution s(w);
    cout << s.pickIndex() << endl;
    cout << s.pickIndex() << endl;
    cout << s.pickIndex() << endl;
    return 0;
}
